src/futures.rs

```diff
@@ -27,3 +27,23 @@ impl<T, F: Fn() -> Option<T>> Future for WaitForValue<F> {
         }
     }
 }
+
+/// Generates a synchronous `main()` function that calls the provided
+/// asynchronous function. To avoid name conflicts, the asynchronous function
+/// must not be called `main()`; `async_main()` is fine.
+///
+/// Example:
+///     libtock::async_main!(async_main);
+///     async fn async_main() {
+///         // Snipped
+///     }
+#[macro_export]
+macro_rules! async_main {
+    ($main_name:ident) => {
+        fn main() {
+            unsafe {
+                ::core::executor::block_on($main_name());
+            }
+        }
+    };
+}
```

Comment:
There is a hidden loophole. You can create a `main` function in any module and access it from safe code.

Reply:
> There is a hidden loophole. You can create a `main` function in any module

Yes, this behavior differs from Rust's normal behavior. However, I don't think it's a particularly big deal -- and in particular, the `async_main!` call must be in the crate root.

> and access it from safe code.

This is true for normal Rust as well.

Reply:
My point was that `block_on` is unsafe. You should only call it within the main loop.

You now *could* call `main` from a callback which we considered `unsafe`. That's what the whole motivation for the `async main` was in the first place.

Reply:
Ah, I see now. Weirdly enough, in a `#[no_std]` context, `main` can be `unsafe`. I changed that locally; I'll push it when I push the fixed examples.

Reply:
I was unable to add a fix that catches this at compile time. Instead, I added a runtime check.